simpleclient/src/main/java/io/prometheus/client/CollectorRegistry.java

```diff
@@ -125,31 +125,53 @@ public void clear() {
   public Enumeration<Collector.MetricFamilySamples> metricFamilySamples() {
     return new MetricFamilySamplesEnumeration();
   }
+
+  public Enumeration<Collector.MetricFamilySamples> metricFamilySamples(Set<String> includedNames) {
+    return new MetricFamilySamplesEnumeration(includedNames);
+  }
+
   class MetricFamilySamplesEnumeration implements Enumeration<Collector.MetricFamilySamples> {
 
     private final Iterator<Collector> collectorIter = collectors().iterator();
     private Iterator<Collector.MetricFamilySamples> metricFamilySamples;
     private Collector.MetricFamilySamples next;
+    private Set<String> includedNames;
 
-    MetricFamilySamplesEnumeration() {
+    MetricFamilySamplesEnumeration(Set<String> includedNames) {
+      this.includedNames = includedNames;
       findNextElement();
     }
 
+    MetricFamilySamplesEnumeration() {
+      this(Collections.<String>emptySet());
+    }
+
     private void findNextElement() {
       if (metricFamilySamples != null && metricFamilySamples.hasNext()) {
-        next = metricFamilySamples.next();
+        next = filter(metricFamilySamples.next());
       } else {
         while (collectorIter.hasNext()) {
           metricFamilySamples = collectorIter.next().collect().iterator();
```

Comment:
So I guess we could leverage the `Describable` interface to determine if we can skip the collector? If the collector doesn't implement `Describable` we can just assume it need to be called and filter it after that.

Or can we add to the collector signature so collectors can be aware of filtering? This seems like a potentially more performant approach, but then it puts the weight on individual collectors and that seems like it is leaking the logic.